src/graph/network.cc

```diff
@@ -0,0 +1,124 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file graph/network.cc
+ * \brief DGL networking related APIs
+ */
+
+#include "./network.h"
+#include "./network/communicator.h"
+#include "./network/socket_communicator.h"
+#include "./network/serialize.h"
+
+#include "../c_api_common.h"
+
+using dgl::runtime::DGLArgs;
+using dgl::runtime::DGLArgValue;
+using dgl::runtime::DGLRetValue;
+using dgl::runtime::PackedFunc;
+using dgl::runtime::NDArray;
+
+namespace dgl {
+namespace network {
+
+static char* sender_data_buffer = nullptr;
+static char* recv_data_buffer = nullptr;
+
+///////////////////////// Distrobuted Sampler /////////////////////////
+
+DGL_REGISTER_GLOBAL("network._CAPI_DGLSenderCreate")
+.set_body([] (DGLArgs args, DGLRetValue* rv) {
+    std::string ip = args[0];
+    int port = args[1];
+    network::Communicator* comm = new network::SocketCommunicator();
+    if (comm->Initialize(IS_SENDER, ip.c_str(), port) == false) {
+      LOG(FATAL) << "Initialize network communicator (sender) error.";
+    }
+    try {
+      sender_data_buffer = new char[kMaxBufferSize];
+    } catch (const std::bad_alloc&) {
+      LOG(FATAL) << "Not enough memory for sender buffer: " << kMaxBufferSize;
+    }
+    CommunicatorHandle chandle = static_cast<CommunicatorHandle>(comm);
+    *rv = chandle;
+  });
+
+DGL_REGISTER_GLOBAL("network._CAPI_DGLReceiverCreate")
+.set_body([] (DGLArgs args, DGLRetValue* rv) {
+    std::string ip = args[0];
+    int port = args[1];
+    int num_sender = args[2];
+    network::Communicator* comm = new network::SocketCommunicator();
+    if (comm->Initialize(IS_RECEIVER, ip.c_str(), port, num_sender, kQueueSize) == false) {
+      LOG(FATAL) << "Initialize network communicator (receiver) error.";
+    }
+    try {
+      recv_data_buffer = new char[kMaxBufferSize];
+    } catch (const std::bad_alloc&) {
+      LOG(FATAL) << "Not enough memory for receiver buffer: " << kMaxBufferSize;
+    }
+    CommunicatorHandle chandle = static_cast<CommunicatorHandle>(comm);
+    *rv = chandle;
+  });
+
+DGL_REGISTER_GLOBAL("network._CAPI_SenderSendSubgraph")
+.set_body([] (DGLArgs args, DGLRetValue* rv) {
+    CommunicatorHandle chandle = args[0];
+    GraphHandle ghandle = args[1];
+    const IdArray node_mapping = IdArray::FromDLPack(CreateTmpDLManagedTensor(args[2]));
+    const IdArray edge_mapping = IdArray::FromDLPack(CreateTmpDLManagedTensor(args[3]));
+    const IdArray layer_offsets = IdArray::FromDLPack(CreateTmpDLManagedTensor(args[4]));
+    const IdArray flow_offsets = IdArray::FromDLPack(CreateTmpDLManagedTensor(args[5]));
+    ImmutableGraph *ptr = static_cast<ImmutableGraph*>(ghandle);
+    network::Communicator* comm = static_cast<network::Communicator*>(chandle);
+    auto csr = ptr->GetInCSR();
```

Comment:
I think it's better to check out csr as well. Although NodeFlow probably has only in csr, we should add a check for now.

Reply:
You mean that we can just serialize an empty out-csr for now?

Reply:
i guess just double check if GetOutCSR() == nll.